FirebaseMLModelDownloader/Sources/ModelInfoRetriever.swift

```diff
@@ -129,82 +108,92 @@ extension ModelInfoRetriever {
     return request
   }
 
-  /// Get model info from server.
-  func downloadModelInfo(completion: @escaping (DownloadError?) -> Void) {
+  /// Get installations auth token.
+  func getAuthToken(completion: @escaping (Result<String, DownloadError>) -> Void) {
     /// Get FIS token.
-    installations.authToken { [weak self] tokenResult, error in
-      guard let self = self else {
-        completion(.internalError(description: ModelInfoRetriever.selfDeallocatedErrorDescription))
-        return
-      }
+    installations.authToken { tokenResult, error in
       guard let result = tokenResult
       else {
-        completion(.internalError(description: ModelInfoRetriever.tokenErrorDescription))
+        completion(.failure(.internalError(description: ModelInfoRetriever.tokenErrorDescription)))
         return
       }
-      /// Get model info fetch URL with appropriate HTTP headers.
-      let request = self.getModelInfoFetchURLRequest(token: result.authToken)
-
-      /// Download model info.
-      // TODO: Consider moving request to a separate method
-      let session = URLSession(configuration: .ephemeral)
-      let dataTask = session.dataTask(with: request) { [weak self]
-        data, response, error in
-        guard let self = self else {
-          completion(.internalError(description: ModelInfoRetriever
-              .selfDeallocatedErrorDescription))
-          return
-        }
-        if let downloadError = error {
-          completion(.internalError(description: downloadError.localizedDescription))
-        } else {
-          guard let httpResponse = response as? HTTPURLResponse else {
+      completion(.success(result.authToken))
+    }
+  }
+
+  /// Get model info from server.
+  func downloadModelInfo(completion: @escaping (DownloadError?) -> Void) {
+    getAuthToken { result in
+      switch result {
+      case let .success(authToken):
+        /// Get model info fetch URL with appropriate HTTP headers.
+        let request = self.getModelInfoFetchURLRequest(token: authToken)
+        let session = URLSession(configuration: .ephemeral)
```

Comment:
A point to consider in the future. Ephemeral session doesn't handle ETag headers, so the model will be downloaded each time even if a local copy has already been downloaded. To optimize it we should either send Etag and handle the response ourself (like Remote Config does) or request consultation with security experts to double check if caching FIS token is actually an issue (if not we may put the shared session back). I think it is worth adding a TODO to consider later.

Reply:
Great point! I've manually set the caching policy for the request for now. However, I have also added a TODO so we can consult with the experts on what the right approach is.